### PR TITLE
Implement unmanaged-to-managed direction for vtable stub generator

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Implement the unmanaged-to-managed direction for stubs generated with the VTableIndexStubGenerator.

This PR also includes a design doc for how to handle exception handling at the unmanaged-to-managed boundary, which this PR implements.

<details>
<summary>Generated Code (NativeToManagedStubs.g.cs)</summary>

```csharp
// <auto-generated/>
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    [System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute]
                    internal static int ABI_GetData(void* __this_native)
                    {
                        global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject @this;
                        int __retVal = default;
                        // Unmarshal - Convert native data to managed data.
                        @this = System.Runtime.InteropServices.IUnmanagedVirtualMethodTableProvider<byte>.GetObjectForUnmanagedWrapper<global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject>(__this_native);
                        __retVal = @this.GetData();
                        return __retVal;
                    }
                }
            }
        }
    }
}
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    [System.Runtime.InteropServices.UnmanagedCallersOnlyAttribute]
                    internal static void ABI_SetData(void* __this_native, int x)
                    {
                        global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject @this;
                        // Unmarshal - Convert native data to managed data.
                        @this = System.Runtime.InteropServices.IUnmanagedVirtualMethodTableProvider<byte>.GetObjectForUnmanagedWrapper<global::ComInterfaceGenerator.Tests.NativeExportsNE.ImplicitThis.INativeObject>(__this_native);
                        @this.SetData(x);
                    }
                }
            }
        }
    }
}

```

</details>

<details>
<summary>Generated code (PopulateVTable.g.cs)</summary>

```csharp
// <auto-generated/>
namespace ComInterfaceGenerator.Tests
{
    internal unsafe partial class NativeExportsNE
    {
        internal unsafe partial class ImplicitThis
        {
            internal unsafe partial interface INativeObject
            {
                internal unsafe partial interface Native
                {
                    internal static void PopulateUnmanagedVirtualMethodTable(System.Span<nint> vtable)
                    {
                        vtable[0] = (nint)(delegate* unmanaged<void*, int> )&ABI_GetData;
                        vtable[1] = (nint)(delegate* unmanaged<void*, int, void> )&ABI_SetData;
                    }
                }
            }
        }
    }
}

```

</details>

### CODE REVIEW

**Review by AaronRobinsonMSFT — COMMENTED** [review]
Can you share an example of what the generated source looks like?

**Comment by jkoritzinsky** [comment]
Once I've done more work and gotten this PR to at least some semblance of green, I'll add some example generated code to the top of this PR.

**Comment by jkoritzinsky** [comment]
I've pushed all of the changes to this PR, but I'm going to spin off some of these commits into separate PRs to make them more reviewable. I'll mark this as ready for review once those PRs are merged.

File: docs/design/libraries/ComInterfaceGenerator/UnmanagedToManagedEH.md
Comment (AaronRobinsonMSFT) [inline docs/design/libraries/ComInterfaceGenerator/UnmanagedToManagedEH.md:3]:
```suggestion
As part of building out the "vtable method stub" generator as the first steps on our path to the COM source generator, we need to determine how to handle exceptions at the managed code boundary in the unmanaged-calling-managed direction. When implementing a COM Callable Wrapper (or equivalent concept), the source generator will need to generate an `[UnmanagedCallersOnly]` wrapper of a method. We do not support propagating exceptions across an `[UnmanagedCallersOnly]` method on all platforms, so we need to determine how to handle exceptions.
```

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeThisToManagedThisMarshallerFactory.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeThisToManagedThisMarshallerFactory.cs:14]:
```suggestion
    internal sealed class NativeThisToManagedThisMarshallerFactory : IMarshallingGeneratorFactory
```

I assume this should match the name of the file.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/UnmanagedToManagedStubGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/UnmanagedToManagedStubGenerator.cs:102]:
```suggestion
            List<StatementSyntax> finallyStatements = new();
```

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionDefaultMarshaller.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionDefaultMarshaller.cs:24]:
```suggestion
            // and calls SetErrorInfo if the platform supports it.
```

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionDefaultMarshaller.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionDefaultMarshaller.cs:26]: I would defer to the runtime at present.

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs:27]: I don't understand this `CreateTruncating` call.

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionNaNMarshaller.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionNaNMarshaller.cs:24]:
```suggestion
            // and calls SetErrorInfo if the platform supports it.
```

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/SwallowExceptionMarshaller.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/SwallowExceptionMarshaller.cs:22]:
```suggestion
            // and calls SetErrorInfo if the platform supports it.
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs:37]: Is this going to be a user error or some internal logic issue? If user error I think we should fail hard.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs:59]: Let's comment on what this and the branch are doing a bit more. They seem very very close but one is computed using some expression matching whereas the below is using some boolean logic. Is there a reason? I assume one is for unmanaged => managed and the other managed => managed judging by the "paramMarshallers" collections.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs:91]: Is this user visible? I can't recall if we are localizing these errors.

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs:30]: It does help, but this is an odd way to express this logic. From what I am seeing here, this code is also applied to `double` and `float`, which means this is no longer an `HRESULT`. I'm also not sure consuming `INumber` here is a good idea as I think it will root all `INumber` interfaces for a COM scenario regardless if this is the only place it is used. Why is the generic needed for `HRESULT`?

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs:30]: I'm using the generic math here to deduplicate a little code. The `ExceptionHResultMarshaller<T>` is meant to cover the cases where `int` and `uint` are used as the unmanaged signature return type. For `ExceptionNaNMarshaller<T>` is meant to cover `float` and `double`. I can switch from the generic math to separate types, but I was trying to avoid code duplication now that we have these interfaces available.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/BoundGenerators.cs:37]: This would be a logic error in the usage of the APIs, not a user error.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorExtensions.cs:91]: This is not user-visible. We generally haven't been localizing these errors.

File: src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportAnalyzer.cs]: I've taken another stab at the shape for `BoundGenerators`. Let me know what you think!

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs:78]: I do like this more, but I'm not sure why having a property is better than having an out on a factory method. Is there a reason you prefer the `new` allocation rather a `BoundGenerators.Create()`? The errors could be an `out` and be clear that they exist.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:50]: This function is incredibly dense and I've little to no idea on what it is trying to do. Can we get it commented?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:105]: The `Select` and `Where` are chained above, is there a reason we don't do that there?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:50]: Upon walking through this I get it, but I've no confidence I could write it. Up to you if it needs any more clarity.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:226]: Comment says "only contains primitive values", but we only check for `int`. Is this accurate?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:347]: Why "1"?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:421]: I really would prefer not to see escaped keywords in our generated code. Let's go with `inst` or `thisPtr` or some other mechanism. We shouldn't need to rely on escaped keywords.

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:421]: I was worried that `inst` or `thisPtr` was more likely to conflict with user-defined identifiers than `@this`, so I went with `@this` so I could avoid having to add the extra underscores and make the code look a little cleaner. I'll change it.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:347]: The definition of `IUnmanagedInterfaceType<>` changed and now the type key is the second type argument instead of the first. I'll add a comment. I don't think it's worth trying to future proof by checking for the type parameter based on name as the type parameter list is going to be unchangable after we ship anyway.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:226]: Yes, this is accurate. Basically `typedConstant.Value is EnumType` will always be `false`, even though you could cast `(EnumType)typedConstant.Value` if `typedConstant.Value` is an `int`.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs:78]: I'll take another shot at trying to move this into a good factory design.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:421]: > I went with @this so I could avoid having to add the extra underscores and make the code look a little cleaner

I see. That is an annoying problem. Sigh. I'll defer to you here, if this makes the codegen easier with less ceremony let's go with it.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs:78]: It is merely a question about design. If you have a strong desire to avoid a `static` factory method, then I'm okay with that. Based on my rudimentary understanding here it seems the factory pattern is most appropriate, but I can be convinced.

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/IUnmanagedVirtualMethodTableProvider.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/IUnmanagedVirtualMethodTableProvider.cs:65]: Add full docs for all the APIs in here (`Ancillary.Interop`)? Probably also be nice for future-us when proposing/adding the APIs.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ManagedToNativeVTableMethodGenerator.cs:78]: I don't have a particularly strong desire against moving this to a factory method, and I think it might end up being slightly cleaner. I just defaulted to keeping it as a constructor since that was an easier change.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:105]: Yes, we use the results of this filter in 2 places instead of just 1, so I put it in a local so we could reuse it.

File: src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs:35]: I think this is the last query I have.

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/Ancillary.Interop/ExceptionHResultMarshaller.cs:35]: I think the tradeoff is worth it. We can explore breaking it up and duplicating the APIs when we take these to API review.

**Review by AaronRobinsonMSFT — APPROVED** [review]
Please give @elinor-fung a chance to review this, but I am good.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypePositionInfo.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypePositionInfo.cs:54]: Should this also check for `ExceptionIndex`?

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypePositionInfo.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypePositionInfo.cs:50]: I might be missing it - where does the generator create a `TypePositionInfo` using `ExceptionIndex` or `IncrementalStubGenerationContext.ExceptionMarshallingInfo`?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:363]: Should we also have a check/diagnostic for invalid `ExceptionMarshalling` + `ExceptionMarshallingCustomType` combination?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeToManagedThisMarshallerFactory.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/NativeToManagedThisMarshallerFactory.cs:41]:
```suggestion
```

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:52]:
```suggestion
            // Get all methods with the [VirtualMethodIndex] attribute.
```

**Review by elinor-fung — APPROVED** [review]
I think it would still be helpful to put examples of the generated source in the PR description. Beyond the PR itself, I have found it helpful when looking through previous changes when reworking (or breaking) things when we were working on LibraryImport.

I think we also need tests that would have caught the exception marshalling wiring (fine with that in a separate change) .

**Comment by jkoritzinsky** [comment]
I'll add the runtime tests for ExceptionMarshalling in a separate change.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs
Comment (elinor-fung) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/VtableIndexStubGenerator.cs:385]: This means we don't respect `ExceptionMarshallingCustomType` if `ExceptionMarshalling` is not specified, right? Is that intentional? I think that would be inconsistent the doc and what we did for `StringMarshalling[CustomType]`.

**Review by elinor-fung — APPROVED** [review]
Thanks for adding those generated code examples!

**Comment by jkoritzinsky** [comment]
Failure is unrelated.

**Pull request merged**
